internal/users/manager.go

```diff
@@ -292,33 +383,254 @@ func (m *Manager) AllShadows() ([]ShadowEntry, error) {
+		// that we already registered (if they do, there's nothing we can do about it).
+		var tmpName string
+		tmpName, cleanup, err = m.addTemporaryGroup(gid)
+		if errors.Is(err, errGroupAlreadyExists) {
+			log.Debugf(context.Background(), "GID %d already in use, generating a new one", gid)
+			continue
+		}
+		if err != nil {
+			return 0, nil, fmt.Errorf("could not register temporary group: %w", err)
+		}
+
+		if unique := m.checkIsUniqueGID(gid, tmpName); unique {
+			break
+		}
+
+		// If the GID is not unique, remove the temporary group and generate a new one in the next iteration.
+		if err := cleanup(); err != nil {
+			return 0, nil, fmt.Errorf("could not remove temporary group %q: %w", tmpName, err)
+		}
+	}
+
+	log.Debugf(context.Background(), "Registered group %q with GID %d", name, gid)
+
+	return gid, cleanup, nil
+}
 
-	// Convert the first 4 bytes of the hash into an integer
-	number := binary.BigEndian.Uint32(hash[:4]) % (maxID + 1)
+func (m *Manager) checkIsUniqueGID(gid uint32, tmpName string) bool {
+	for _, entry := range localgroups.GetGroupEntries() {
+		if entry.GID == gid && entry.Name != tmpName {
+			log.Debugf(context.Background(), "GID %d already in use by group %q, generating a new one", gid, entry.Name)
+			return false
+		}
+	}
 
-	// Repeat hashing until we get a number in the desired range. This ensures that the generated IDs are uniformly
-	// distributed in the range, opposed to a simple modulo operation.
-	for number < minID {
-		hash = sha256.Sum256(hash[:])
-		number = binary.BigEndian.Uint32(hash[:4]) % (maxID + 1)
+	return true
+}
+
+var errGroupAlreadyExists = errors.New("group already exists")
+
+func (m *Manager) addTemporaryGroup(gid uint32) (name string, cleanup func() error, err error) {
+	// Check if the GID is already registered.
+	_, err = m.GroupByID(gid)
+	if err == nil {
+		return "", nil, errGroupAlreadyExists
+	}
+	if !errors.Is(err, NoDataFoundError{}) {
+		return "", nil, err
+	}
+
+	// Generate a 32 character (16 bytes in hex) random name.
+	bytes := make([]byte, 16)
+	if _, err := rand.Read(bytes); err != nil {
+		return "", nil, fmt.Errorf("failed to generate random name: %w", err)
+	}
+	name = fmt.Sprintf("%x", bytes)
+
+	m.temporaryGroups[gid] = name
+
+	cleanup = func() error {
+		return m.deleteTemporaryGroup(gid)
+	}
+
+	return name, cleanup, nil
+}
+
+func (m *Manager) deleteTemporaryGroup(gid uint32) error {
+	_, ok := m.temporaryGroups[gid]
+	if !ok {
+		return fmt.Errorf("temporary group with GID %d does not exist", gid)
+	}
+
+	delete(m.temporaryGroups, gid)
+	return nil
+}
+
+// GenerateUID generates a random UID in the configured range.
+func (m *Manager) GenerateUID() (uint32, error) {
+	if m.uidsToGenerateInTests != nil {
+		if len(m.uidsToGenerateInTests) == 0 {
+			return 0, fmt.Errorf("no more UIDs to generate in tests")
+		}
+
+		uid := m.uidsToGenerateInTests[0]
+		m.uidsToGenerateInTests = m.uidsToGenerateInTests[1:]
+		return uid, nil
+	}
```

Comment:
So instead of this, it would be nice if we could overwrite the behavior at test runtime only